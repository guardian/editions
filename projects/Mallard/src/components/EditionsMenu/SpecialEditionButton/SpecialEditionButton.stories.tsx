import React from 'react'
import { storiesOf } from '@storybook/react-native'
import { withKnobs, text } from '@storybook/addon-knobs'
import { SpecialEditionButton } from './SpecialEditionButton'

const props = {
    devUri:
        'https://media.guim.co.uk/49cebb0db4a3e4d26d7d190da7be4a2e9bd7534f/0_0_103_158/103.png',
    expiry: new Date(),
    image: {
        source: '',
        path: '',
    },
    onPress: () => {},
    title: text(
        'Title',
        `Food
Monthly`,
    ),
    subTitle: text(
        'SubTitle',
        'Store cupboard special: 20 quick and easy lockdown suppers',
    ),
    style: {
        backgroundColor: '#FEEEF7',
        expiry: {
            color: '#7D0068',
            font: 'GuardianTextSans-Regular',
            lineHeight: 16,
            size: 15,
        },

        subTitle: {
            color: '#7D0068',
            font: 'GuardianTextSans-Bold',
            lineHeight: 20,
            size: 17,
        },
        title: {
            color: '#121212',
            font: 'GHGuardianHeadline-Regular',
            lineHeight: 34,
            size: 34,
        },
        image: {
            height: 134,
            width: 87,
        },
    },
}

storiesOf('SpecialEditionButton', module)
    .addDecorator(withKnobs)
    // This is useful to check the default style but breaks TS
    // .add('SpecialEditionButton - default', () => (
    //     <SpecialEditionButton {...props} style={null} />
    // ))
    .add('SpecialEditionButton - with styles', () => (
        <SpecialEditionButton {...props} />
    ))
    .add('SpecialEditionButton - selected', () => (
        <SpecialEditionButton selected {...props} />
    ))
