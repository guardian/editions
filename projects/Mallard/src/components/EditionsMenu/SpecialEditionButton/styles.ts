import { color } from 'src/theme/color'
import {
    TextFormatting,
    SpecialEditionButtonStyles,
} from '../../../../../Apps/common/src'
import { StyleSheet } from 'react-native'
import { EDITIONS_MENU_TEXT_LEFT_PADDING } from '../EditionsMenu'

const expiryDefaults = {
    color: color.text,
    font: 'GuardianTextSans-Regular',
    lineHeight: 16,
    size: 15,
}

const titleDefaults = {
    color: color.text,
    font: 'GHGuardianHeadline-Regular',
    lineHeight: 34,
    size: 34,
}

const subTitleDefaults = {
    color: color.text,
    font: 'GuardianTextSans-Bold',
    lineHeight: 20,
    size: 17,
}

const textFormatting = (
    selected: boolean,
    style: TextFormatting,
    defaults: TextFormatting,
) => ({
    color: selected ? 'white' : (style && style.color) || defaults.color,
    fontFamily: (style && style.font) || defaults.font,
    fontSize: (style && style.size) || defaults.size,
    lineHeight: (style && style.lineHeight) || defaults.lineHeight,
})

const styles = ({
    style,
    selected,
}: {
    style: SpecialEditionButtonStyles
    selected: boolean
}) => {
    const imageWidth = (style && style.image && style.image.width) || 67
    return StyleSheet.create({
        container: {
            backgroundColor:
                (selected ? color.primary : style && style.backgroundColor) ||
                color.palette.neutral[97],
            flexDirection: 'row',
            paddingTop: 12,
        },
        expiry: {
            flexWrap: 'wrap',
            marginTop: 8,
            ...textFormatting(selected, style && style.expiry, expiryDefaults),
        },
        image: {
            width: imageWidth,
            height: (style && style.image && style.image.height) || 134,
        },
        textBox: {
            flexShrink: 1,
            paddingLeft: EDITIONS_MENU_TEXT_LEFT_PADDING - imageWidth,
            paddingBottom: 15,
            paddingRight: 20,
        },
        title: {
            flexWrap: 'wrap',

            ...textFormatting(selected, style && style.title, titleDefaults),
        },
        subTitle: {
            flexWrap: 'wrap',
            marginTop: 10,
            ...textFormatting(
                selected,
                style && style.subTitle,
                subTitleDefaults,
            ),
        },
    })
}
export { styles }
