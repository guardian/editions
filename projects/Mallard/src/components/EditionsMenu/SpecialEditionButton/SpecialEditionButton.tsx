import moment from 'moment'
import React, { useState } from 'react'
import { Text, TouchableWithoutFeedback, View } from 'react-native'
import { ImageResource } from 'src/components/front/image-resource'
import {
    Image,
    SpecialEditionButtonStyles,
} from '../../../../../Apps/common/src'
import { styles } from './styles'

const SpecialEditionButton = ({
    expiry,
    devUri,
    image,
    onPress,
    selected = false,
    subTitle,
    style,
    title,
}: {
    expiry: Date
    devUri?: string
    image: Image
    onPress: () => void
    selected?: boolean
    style: SpecialEditionButtonStyles
    subTitle: string
    title: string
}) => {
    const [pressed, setPressed] = useState(false)
    const defaultStyles = styles({ style, selected: selected || pressed })

    return (
        <TouchableWithoutFeedback
            accessibilityRole="button"
            accessibilityLabel={`${title} - Special Edition Button`}
            onPressIn={() => setPressed(true)}
            onPressOut={() => setPressed(false)}
            onPress={onPress}
        >
            <View style={defaultStyles.container}>
                <ImageResource
                    devUri={devUri}
                    image={image}
                    use="full-size"
                    style={defaultStyles.image}
                />
                <View style={defaultStyles.textBox}>
                    <Text style={defaultStyles.title}>{title}</Text>
                    <Text style={defaultStyles.subTitle}>{subTitle}</Text>
                    <Text style={defaultStyles.expiry}>
                        Available until{' '}
                        {moment(expiry)
                            .local()
                            .format('l')}
                    </Text>
                </View>
            </View>
        </TouchableWithoutFeedback>
    )
}

export { SpecialEditionButton }
