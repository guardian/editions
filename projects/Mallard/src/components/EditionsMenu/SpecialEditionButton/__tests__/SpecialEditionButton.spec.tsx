import React from 'react'
import TestRenderer, { ReactTestRendererJSON } from 'react-test-renderer'
import { SpecialEditionButton } from '../SpecialEditionButton'

jest.mock('src/components/front/image-resource', () => ({
    ImageResource: () => 'ImageResource',
}))

const fixtures = {
    expiry: new Date(98, 1),
    image: {
        source: 'media',
        path: '/path/to/image',
    },
    onPress: () => {},
    title: `Food
Monthly`,
    subTitle: 'Store cupboard special: 20 quick and easy lockdown suppers',
    style: {
        backgroundColor: '#FEEEF7',
        expiry: {
            color: '#7D0068',
            font: 'GuardianTextSans-Regular',
            lineHeight: 16,
            size: 15,
        },

        subTitle: {
            color: '#7D0068',
            font: 'GuardianTextSans-Bold',
            lineHeight: 20,
            size: 17,
        },
        title: {
            color: '#121212',
            font: 'GHGuardianHeadline-Regular',
            lineHeight: 34,
            size: 34,
        },
        image: {
            height: 134,
            width: 87,
        },
    },
}

export { fixtures }

describe('SpecialEditionButton', () => {
    it('should display a default SpecialEditionButton with imported styling', () => {
        const component: ReactTestRendererJSON | null = TestRenderer.create(
            <SpecialEditionButton {...fixtures} />,
        ).toJSON()
        expect(component).toMatchSnapshot()
    })
    it('should display a selected SpecialEditionButton despite imported styling', () => {
        const component: ReactTestRendererJSON | null = TestRenderer.create(
            <SpecialEditionButton selected {...fixtures} />,
        ).toJSON()
        expect(component).toMatchSnapshot()
    })
})
