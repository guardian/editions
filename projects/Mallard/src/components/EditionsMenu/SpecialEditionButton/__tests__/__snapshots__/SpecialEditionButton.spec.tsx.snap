// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpecialEditionButton should display a default SpecialEditionButton with imported styling 1`] = `
<View
  accessibilityLabel="Food
Monthly - Special Edition Button"
  accessibilityRole="button"
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#FEEEF7",
      "flexDirection": "row",
      "paddingTop": 12,
    }
  }
>
  <Image
    resizeMethod="resize"
    source={
      Object {
        "uri": "https://media.guim.co.uk/49cebb0db4a3e4d26d7d190da7be4a2e9bd7534f/0_0_103_158/103.png",
      }
    }
    style={
      Object {
        "height": 134,
        "width": 87,
      }
    }
  />
  <View
    style={
      Object {
        "flexShrink": 1,
        "paddingBottom": 15,
        "paddingLeft": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#121212",
          "flexWrap": "wrap",
          "fontFamily": "GHGuardianHeadline-Regular",
          "fontSize": 34,
          "lineHeight": 34,
        }
      }
    >
      Food
Monthly
    </Text>
    <Text
      style={
        Object {
          "color": "#7D0068",
          "flexWrap": "wrap",
          "fontFamily": "GuardianTextSans-Bold",
          "fontSize": 17,
          "lineHeight": 20,
          "marginTop": 10,
        }
      }
    >
      Store cupboard special: 20 quick and easy lockdown suppers
    </Text>
    <Text
      style={
        Object {
          "color": "#7D0068",
          "flexWrap": "wrap",
          "fontFamily": "GuardianTextSans-Regular",
          "fontSize": 15,
          "lineHeight": 16,
          "marginTop": 8,
        }
      }
    >
      Available until 
      2/1/1998
    </Text>
  </View>
</View>
`;

exports[`SpecialEditionButton should display a selected SpecialEditionButton despite imported styling 1`] = `
<View
  accessibilityLabel="Food
Monthly - Special Edition Button"
  accessibilityRole="button"
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#052962",
      "flexDirection": "row",
      "paddingTop": 12,
    }
  }
>
  <Image
    resizeMethod="resize"
    source={
      Object {
        "uri": "https://media.guim.co.uk/49cebb0db4a3e4d26d7d190da7be4a2e9bd7534f/0_0_103_158/103.png",
      }
    }
    style={
      Object {
        "height": 134,
        "width": 87,
      }
    }
  />
  <View
    style={
      Object {
        "flexShrink": 1,
        "paddingBottom": 15,
        "paddingLeft": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "flexWrap": "wrap",
          "fontFamily": "GHGuardianHeadline-Regular",
          "fontSize": 34,
          "lineHeight": 34,
        }
      }
    >
      Food
Monthly
    </Text>
    <Text
      style={
        Object {
          "color": "white",
          "flexWrap": "wrap",
          "fontFamily": "GuardianTextSans-Bold",
          "fontSize": 17,
          "lineHeight": 20,
          "marginTop": 10,
        }
      }
    >
      Store cupboard special: 20 quick and easy lockdown suppers
    </Text>
    <Text
      style={
        Object {
          "color": "white",
          "flexWrap": "wrap",
          "fontFamily": "GuardianTextSans-Regular",
          "fontSize": 15,
          "lineHeight": 16,
          "marginTop": 8,
        }
      }
    >
      Available until 
      2/1/1998
    </Text>
  </View>
</View>
`;
